File: LEETCODE/MEDIUM/DAY 31/PYTHON SOLUTIONS/BOTTOM-UP RECURSIVE.py
class Solution:
    def upsideDownBinaryTree(self, root: Optional[TreeNode], parent: Optional[TreeNode] = None) -> Optional[TreeNode]:
        if not root:
            return parent
        newRoot = self.upsideDownBinaryTree(root.left, root)
        root.left = parent.right if parent else None
        root.right = parent
        return newRoot

File: LEETCODE/MEDIUM/DAY 31/JAVA SOLUTIONS/BOTTOM-UP RECURSIVE.java
public TreeNode UpsideDownBinaryTree(TreeNode root) {
    return dfsBottomUp(root, null);
}

private TreeNode dfsBottomUp(TreeNode p, TreeNode parent) {
    if (p == null) return parent;
    TreeNode newRoot = dfsBottomUp(p.left, p);
    p.left = (parent == null) ? null : parent.right;
    p.right = parent;
    return newRoot;
}
